Add setup wizard data manager

renatho opened the pull request:
### Changes proposed in this Pull Request

* Talking with @yscik, we decided to introduce a data manager to the setup wizard, using the `@wordpress/data`. It's a better approach because we're handling so much data and requests in the setup wizard and a data store will help us with some features.
* As a first case, I refactored the Welcome step data.
* The idea is to remove the useOnboardingApi in favor of the `@wordpress/data` store.

### Testing instructions

* Go to `/wp-admin/admin.php?page=sensei_onboarding&step=welcome`
* Click on "Continue"
* Toggle the select and click on "Continue"
* Go to `/wp-admin/admin.php?page=sensei-settings` and make sure the `Enable usage tracking` is updated.

renatho commented on assets/onboarding/welcome/index.jsx:
```diff
@@ -14,13 +14,27 @@ export const Welcome = () => {
+	);
+	const { submitUsageTracking } = useDispatch( 'sensei-setup-wizard' );
+
+	const isBusy = isFetching || isSubmitting;
```
I separated the `isFetching` and `isSubmitting`, so we can change the loading feedback for the user as we desire. 

Maybe while we are fetching (including the other screens), we could add a loading in the Stepper (this is prepared to receive a loading). - When we work in the navigation persistency we could check this.

renatho commented on assets/onboarding/data/reducer.js:
```diff
@@ -0,0 +1,76 @@
+			};
+
+		case ERROR_USAGE_TRACKING:
+			return {
```
We are not treating error while saving. I added it to the store, so if we desire to show some error message or something like that, we can.

yscik commented:
Looking good!

I think we can simplify the API by having just one endpoint to fetch state for all steps. It's not needed right now to only load a step's data when we open it, as the steps don't really affect each other, and it would speed things up significantly. 

Along this line, it would be fine to have one global `isFetching`, `isSubmitting` and error states in the store, and also only one `fetchSetupWizardUserData` action and resolver that loads everything in one request. (The data itself can be still grouped by step.)

We can replace the step content with a Spinner when `isFetching` is going on (which would mostly be upon initial page load), and keep the `isSubmitting` feedback on the submit buttons like we have now.

yscik commented on assets/onboarding/data/actions.js:
```diff
@@ -0,0 +1,92 @@
+ *
+ * @param {boolean} usageTracking Usage tracking.
+ */
+export function* submitUsageTracking( usageTracking ) {
```
It would be better to name these like `completeWelcomeStep`, because they have other effects (creating pages in this case, advancing the step state). It also looks better for a reducer that advances the current step on COMPLETE_*_STEP.

renatho replied:
With the new approach I kept the `submitWelcomeStep` action creator with a similar name, but the action types are with names like:

`START_SUBMIT_SETUP_WIZARD_DATA`, `SUCCESS_SUBMIT_SETUP_WIZARD_DATA`, and `ERROR_SUBMIT_SETUP_WIZARD_DATA`.

I used `SUCCESS` instead of `COMPLETE` because when there is an error, it's also completed.

renatho commented:
> I think we can simplify the API by having just one endpoint to fetch state for all steps. It's not needed right now to only load a step's data when we open it, as the steps don't really affect each other, and it would speed things up significantly.

Good idea! I refactored it here: https://github.com/Automattic/sensei/pull/3157/commits/b91fa919ff45520044d1b6f8447ea7f0738fd944
We need a very small refactory in the [`fetchSetupWizardData` action creator](https://github.com/Automattic/sensei/pull/3157/files#diff-cff34058a11f48bc41466569ce8c7398R35-R42) when the REST API is ready for that too.

> Along this line, it would be fine to have one global `isFetching`, `isSubmitting` and error states in the store, and also only one `fetchSetupWizardUserData` action and resolver that loads everything in one request. (The data itself can be still grouped by step.)

Also refactored here: https://github.com/Automattic/sensei/pull/3157/commits/b91fa919ff45520044d1b6f8447ea7f0738fd944. Now we need to create only an action creator similar to `submitWelcomeStep` for the other submits and another to update the store like the `setWelcomeStepData`.

> We can replace the step content with a Spinner when `isFetching` is going on (which would mostly be upon initial page load), and keep the `isSubmitting` feedback on the submit buttons like we have now.

Good approach! Also refatored here: https://github.com/Automattic/sensei/pull/3157/commits/b91fa919ff45520044d1b6f8447ea7f0738fd944

renatho commented on assets/onboarding/data/reducer.js:
```diff
@@ -0,0 +1,83 @@
+				},
+			};
+
+		case START_SUBMIT_SETUP_WIZARD_DATA:
```
Oops! Great catch! FIxed here: https://github.com/Automattic/sensei/pull/3157/commits/49c2ed59f4c8e3117046c287a5ed3f548ce213d4

renatho commented on assets/onboarding/data/reducer.js:
```diff
@@ -0,0 +1,83 @@
+				...state,
+				isFetching: false,
+				isSubmitting: false,
+				error: false,
```
Good catch! Also fixed here: https://github.com/Automattic/sensei/commit/49c2ed59f4c8e3117046c287a5ed3f548ce213d4

And I took the opportunity to refactor the fetch part to be similar to the submit and also catch the error.
